### PR TITLE
Ensure nftables unit test parity with iptables

### PR DESCRIPTION
**Pull request by npinaeva**
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
https://github.com/kubernetes/kubernetes/issues/122572
Add missing packet tracing unit tests

#### Special notes for your reviewer:
Unit test rules are coming from https://github.com/kubernetes/kubernetes/blob/b98e47dd80166935117f890b437c8e30a91da173/pkg/proxy/nftables/proxier_test.go#L293, but all chain declarations were put in the beginning to avoid "no such chain" error

To ensure removed parsing functions are not used, we can go to https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/nftables/proxier.go and search for "dport" usages.
You will see that there are 2 possible following symbols:
- `vmap`
- `svcInfo.Port()`
so the removed regex's will never be matched

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

### CODE REVIEW

File: pkg/proxy/nftables/helpers_test.go [hunk 1]
```diff
@@ -815,3 +815,196 @@ func Test_diffNFTablesChain(t *testing.T) {
 		t.Errorf("unexpected difference in empty-chain with trailing newline:\n%s", diff)
 	}
 }
+
+// This tests tracePackets against static data, just to make sure we match things in the
```
Comment (tnqn) [inline pkg/proxy/nftables/helpers_test.go:819]: There is no func called `tracePackets`. Since it's copied from iptables test, should we correct both places to `tracePacket`?

File: pkg/proxy/nftables/helpers_test.go [hunk 2]
```diff
@@ -815,3 +815,196 @@ func Test_diffNFTablesChain(t *testing.T) {
...
+		},
+		{
+			name:     "multiple endpoints",
+			sourceIP: "10.0.0.2",
+			destIP:   "172.30.0.44",
+			destPort: 80,
+			output:   "10.180.0.4:80, 10.180.0.5:80",
+		},
+		{
+			name:     "LOCAL, KUBE-MARK-MASQ",
```
Comment (tnqn) [inline pkg/proxy/nftables/helpers_test.go:988]: `KUBE-MARK-MASQ` is an iptables chain but not a nftables chain, should the name just be `LOCAL, MASQUERADE`?

File: go.mod [hunk 3]
```diff
@@ -121,7 +121,7 @@ require (
 	k8s.io/sample-apiserver v0.0.0
 	k8s.io/system-validators v1.8.0
 	k8s.io/utils v0.0.0-20230726121419-3b25d923346b
-	sigs.k8s.io/knftables v0.0.14
+	sigs.k8s.io/knftables v0.0.15
```
Comment (danwinship) [inline go.mod:124]: You should do the knftables bump as a separate commit

File: pkg/proxy/nftables/helpers_test.go (hunk 2 above)
Comment (danwinship) [reply pkg/proxy/nftables/helpers_test.go:988]: "LOCAL" (and "KUBE-MARK-MASQ") here refer to the specific iptables rule that it's testing in the iptables example ("`... --src-type LOCAL -j KUBE-MARK-MASQ`"). As it happens, this exactly corresponds to the "`... fib saddr type local jump mark-for-masquerade`" in this test, so maybe "local, mark-for-masquerade" as a name.

In general, the goal of the test case is to make sure that each of the regexes in `helpers_test.go` and their corresponding `switch` clauses in `runChain` gets tested. So you should confirm that. You may need some slightly different tests than are in the iptables version since the rules don't all work the same way...

**Review by danwinship — COMMENTED** [review]
> but all chain declarations were put in the beginning to avoid "no such chain" error

actually, now that we have `ParseDump`, can you fix the ordering of the dump in `TestOverallNFTablesRules` and then call `ParseDump` on it as part of the unit test, to make sure that people keep the rules in the right order in the future as things change?

(as a separate commit from the new unit test)

File: pkg/proxy/nftables/helpers_test.go (hunk 1 above)
Comment (npinaeva) [reply pkg/proxy/nftables/helpers_test.go:819]: good point, let me fix this

File: pkg/proxy/nftables/helpers_test.go [hunk 4]
```diff
@@ -815,3 +776,321 @@ func Test_diffNFTablesChain(t *testing.T) {
...
+		},
+		{
+			name:     "multiple endpoints",
+			sourceIP: "10.0.0.2",
+			destIP:   "172.30.0.44",
+			destPort: 80,
+			output:   "10.180.0.4:80, 10.180.0.5:80",
+		},
+		{
+			name:     "LOCAL, mark for masquerade",
```
Comment (danwinship) [inline pkg/proxy/nftables/helpers_test.go:939]: "local, mark for masquerade"
("LOCAL" in all caps refers to part of the iptables rule, but the corresponding nftables rule uses "local")

File: pkg/proxy/nftables/helpers_test.go [hunk 5]
```diff
@@ -815,3 +776,321 @@ func Test_diffNFTablesChain(t *testing.T) {
...
+			sourceIP: testNodeIP,
+			destIP:   "192.168.99.22",
+			destPort: 80,
+			output:   "10.180.0.2:80",
+			masq:     true,
+		},
+		{
+			name:     "DROP",
```
Comment (danwinship) [inline pkg/proxy/nftables/helpers_test.go:947]: (It's fine to keep this in all-caps though since the tracePacket output is still "DROP" even though the nftables rule is "drop").

File: go.mod [hunk 6]
```diff
@@ -120,7 +120,7 @@ require (
 	k8s.io/sample-apiserver v0.0.0
 	k8s.io/system-validators v1.8.0
 	k8s.io/utils v0.0.0-20230726121419-3b25d923346b
-	sigs.k8s.io/knftables v0.0.14
+	sigs.k8s.io/knftables v0.0.16-0.20240228165918-78f68d9b296e
```
Comment (danwinship) [inline go.mod:123]: So it looks like we're good to tag this release then? (Also, this commit should be before the commit that depends on it, not after.)

Comment (danwinship) [reply go.mod:123]: ok, `v0.0.16` is tagged

File: staging/src/k8s.io/kms/internal/plugins/_mock/go.work.sum
[the remark this answers is not in the document]
Comment (npinaeva) [reply staging/src/k8s.io/kms/internal/plugins/_mock/go.work.sum:1]: `make update` did that for some reason, seems to be gone now

**Comment by danwinship** [comment]
/lgtm

/assign @thockin
for vendor update to latest knftables
(which is now pulling in non-code files (README.md, etc) that it didn't used to pull in? Is that expected with the new workspaces-based scripts?)
